doc/langref.html.in

```diff
@@ -2180,16 +2180,23 @@ or
       Unlike normal structs, {#syntax#}packed{#endsyntax#} structs have guaranteed in-memory layout:
       </p>
       <ul>
-        <li>Fields remain in the order declared, least to most significant.</li>
+        <li>Fields are arranged in the order declared, from the least to the most significant bits of a backing integer.</li>
         <li>There is no padding between fields.</li>
-        <li>Zig supports arbitrary width {#link|Integers#} and although normally, integers with fewer
-        than 8 bits will still use 1 byte of memory, in packed structs, they use
-        exactly their bit width.
+        <li>The backing integer has the same bit width as the fields' total bit width.</li>
+        <li>The backing integer is subject to the same rules as any integer, including {#link|alignment|Alignment#}, 
+        having a maximum bit count of 65535, and the host endianness.
+        On a big endian system, the first declared field will have the highest memory address, and on a little endian system, the lowest.
         </li>
-        <li>{#syntax#}bool{#endsyntax#} fields use exactly 1 bit.</li>
+        <li>Due to {#link|alignment|Alignment#}, the backing integer may require more memory that its bit width. {#link|@bitSizeOf|@bitSizeOf#} and {#link|@sizeOf|@sizeOf#} can interrogate the difference.</li>
+        <li>Field access and assignment can be understood as shorthand for bitshifts on the backing integer.</li>
```

Comment:
This one should be left off.  The compiler has many options for implementing packed struct operations, many don't involve bitshifting at all.

How packed structs are translated into machine code is an implementation detail which doesn't need to be analogized here.

Reply:
It is most relevant to the MMIO case, perhaps I could move it down there? See recent comments on this PR about how cpus cannot change individual bits in memory without a load/store of complete bytes. 

I tried to keep it vague (using "can be understood as") because it makes sense that the compiler could optimize field assignments inside very large packed structs to only affect the changed memory, as you said, and because I do not know the exact specifics of the current or expected behavior when volatile is used.

Reply:
The problem I see is that it isn't a complete description. For instance `pack.field3 = true` might be implemented as `backing_int | 0b0100`, that's a mask, not a shift.  The documentation is a user guide: a complete description of the implementation of packed struct operations would be inappropriate, so I don't think a partial analogy is correct here either.

Compare to the [integer literal documentation](https://ziglang.org/documentation/master/#Runtime-Integer-Values). It does say that signed values are always two's complement, that's useful to know: the C standard, for instance, does not (or didn't until quite recently) specify an integer representation. But it doesn't say anything about how the Zig compiler implements large integers like `u256`, because while that's interesting, it isn't user documentation.

As for moving the idea to the MMIO case: yeah that makes perfect sense to me. A comment over a specific operation in the sample code saying "this is equivalent to {stuff}" which shows a mask, bitshift, whatever is correct, that's helpful, because it can be completely correct for the commented line, and illustrates the connection between packed struct field access and the manual-transmission way of doing these things in other low-level languages. So I'd say that's a good place for it.

By saying it's _equivalent_, it's also not a statement about what the implementation happens to do. Keeping implementation details out of the documentation is important, because it's no longer a detail if it's documented.